cloudinit/distros/gentoo.py

```diff
@@ -31,18 +37,33 @@ def __init__(self, name, cfg, paths):
         self.osfamily = "gentoo"
         # Fix sshd restarts
         cfg["ssh_svcname"] = "/etc/init.d/sshd"
+        if distros.uses_systemd():
+            LOG.error("Cloud-init does not support systemd with gentoo")
+
+    def apply_locale(self, _, out_fn=None):
+        """rc-only - not compatible with systemd
+
+        Locales need to be added to /etc/locale.gen and generated prior
+        to selection. Default to en_US.UTF-8 for simplicity.
+        """
+        util.write_file(
+            self.locale_gen_file, "\n".join(self.locales), mode=644
+        )
+
+        # generate locales
+        subp.subp(["locale-gen"], capture=False)
 
-    def apply_locale(self, locale, out_fn=None):
-        if not out_fn:
-            out_fn = self.locale_conf_fn
-        subp.subp(["locale-gen", "-G", locale], capture=False)
+        # select locale
+        subp.subp(
+            ["eselect", "locale", "set", self.default_locale], capture=False
+        )
         # "" provides trailing newline during join
         lines = [
             util.make_header(),
-            'LANG="%s"' % locale,
+            'LANG="%s"' % self.default_locale,
             "",
         ]
-        util.write_file(out_fn, "\n".join(lines))
+        util.write_file(self.locale_conf_fn, "\n".join(lines))
```

Comment:
writing to `/etc/env.d/02locale` is not necessary if running the `eselect locale` as above.  `eselect locale` writes to `/etc/locale.conf` and `/etc/env.d/02locale` is a symlink to `/etc/locale.conf`


```
ls -l /etc/env.d/02locale
lrwxrwxrwx 14 root 2015-02-02 22:36  /etc/env.d/02locale -> ../locale.conf

cat /etc/locale.conf
# Configuration file for eselect
# This file has been automatically generated.
LANG="en_US.utf8"
```

Reply:
It seems I misunderstood what this part was doing, thanks. In that case I guess this might have once succeeded and set the locale. It seems that something must have changed (either in image build or `locale-gen` behavior) in some way if `locale-gen -G en_US.UTF-8` used to succeed.

Either way, I think that this adds some improvement by ensuring that `C.UTF8 UTF-8` is contained in `/etc/locale.gen` before trying to generate the local.